### PR TITLE
feat: add multiple components for account number preferences

### PR DESCRIPTION
**Pull request by luckyman20**
## Description

Functionalities added are - 
1. View all Account Number Preferences
2. Create an Account Number Preference
3. Edit an Account Number Preference
4. Delete an Account Number Preference
5. View an Account Number Preference

Components added are - 
1. account-number-preferences
2. create-account-number-preference
3. edit-account-number-preference
4. view-account-number-preference

## Related issues and discussion
#382

## Screenshots, if any

view all
View all Account Number Preferences

edit account number preferences
Edit Account Number Preference

create anp
Create Account Number Preference

view anp
View an Account Number Preference

## Checklist
Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] If you have multiple commits please combine them into one commit by squashing them.

- [x] Read and understood the contribution guidelines at `web-app/.github/CONTRIBUTING.md`.

### CODE REVIEW

File: src/app/system/account-number-preferences/account-number-preferences.component.scss [hunk 1]
```diff
@@ -0,0 +1,7 @@
+table {
+    width: 100%;
+
+    .select-row:hover {
+      cursor: pointer;
+    }
+  }
```
Comment (abhaychawla) [inline src/app/system/account-number-preferences/account-number-preferences.component.scss]:
```suggestion
}
```
Add a line at the end of file.

File: src/app/system/account-number-preferences/account-number-preferences.component.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+import { MatTableDataSource, MatPaginator, MatSort } from '@angular/material';
+import { ActivatedRoute } from '@angular/router';
+
+@Component({
+  selector: 'mifosx-account-number-preferences',
+  templateUrl: './account-number-preferences.component.html',
+  styleUrls: ['./account-number-preferences.component.scss']
+})
+
+/**
```
Comment (abhaychawla) [inline src/app/system/account-number-preferences/account-number-preferences.component.ts]: Incorrectly placed comment. Correct it at all places.

File: src/app/system/account-number-preferences/account-number-preferences.component.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    this.setAccountNumberPreferences();
+  }
+
+  /**
+   * Initializes the data source, paginator and sorter for account number preferences table.
+   */
+  setAccountNumberPreferences() {
+    this.dataSource = new MatTableDataSource(this.accountNumberPreferencesData);
+    this.dataSource.paginator = this.paginator;
+    this.dataSource.sort = this.sort;
```
Comment (abhaychawla) [inline src/app/system/account-number-preferences/account-number-preferences.component.ts]: Sorting is not working.

File: src/app/system/account-number-preferences/create-account-number-preference/create-account-number-preference.component.html [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+      <mat-card-actions fxLayout="row" fxLayout.xs="column" fxLayoutAlign="center" fxLayoutGap="5px">
+        <button type="button" mat-raised-button [routerLink]="['../']">Cancel</button>
+        <button mat-raised-button color="primary" [disabled]="!accountNumberPreferenceForm.valid">Submit</button>
+      </mat-card-actions>
+
+    </form>
+
+  </mat-card>
+
+</div>
```
Comment (abhaychawla) [inline src/app/system/account-number-preferences/create-account-number-preference/create-account-number-preference.component.html]: Add a line at the end of file.

File: src/app/system/account-number-preferences/create-account-number-preference/create-account-number-preference.component.scss [hunk 5]
```diff
@@ -0,0 +1,4 @@
+.container {
+    max-width: 37rem;
+  }
```
Comment (abhaychawla) [inline src/app/system/account-number-preferences/create-account-number-preference/create-account-number-preference.component.scss]:
```suggestion
}
```

File: src/app/system/account-number-preferences/create-account-number-preference/create-account-number-preference.component.scss [hunk 6]
```diff
@@ -0,0 +1,4 @@
...
+
```
Comment (abhaychawla) [inline src/app/system/account-number-preferences/create-account-number-preference/create-account-number-preference.component.scss]: Ensure only a single line is being added at the end of file.

File: src/app/system/account-number-preferences/create-account-number-preference/create-account-number-preference.component.ts [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    });
+  }
+
+  /**
+   * Creates the account number preference form.
+   * Subscribe on Form Controls to change Prefix Type data.
+   */
+  ngOnInit() {
+    this.createAccountNumberPreferenceForm();
+    this.changePrefixTypeValue();
```
Comment (abhaychawla) [inline src/app/system/account-number-preferences/create-account-number-preference/create-account-number-preference.component.ts]:
```suggestion
    this.getPrefixTypeValue();
```

File: src/app/system/account-number-preferences/create-account-number-preference/create-account-number-preference.component.ts [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+      'prefixType': [{ value: '', disabled: true }]
+    });
+  }
+
+  /**
+   * Submits the account number preference form and creates a account number preference,
+   * if successful redirects to view account number preference.
+   */
+  submit() {
+    this.accountNumberPreferenceForm.get('addPrefix').disable();
```
Comment (abhaychawla) [inline src/app/system/account-number-preferences/create-account-number-preference/create-account-number-preference.component.ts]: Why is this needed?

File: src/app/system/account-number-preferences/create-account-number-preference/create-account-number-preference.component.html [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+        <div fxLayout="column">
+
+          <mat-form-field>
+            <mat-label>Select Account Type</mat-label>
+            <mat-select required formControlName="accountType">
+              <mat-option *ngFor="let accountType of accountNumberPreferencesTemplateData.accountTypeOptions" [value]="accountType.id">
+                {{ accountType.value }}
+              </mat-option>
+            </mat-select>
```
Comment (abhaychawla) [inline src/app/system/account-number-preferences/create-account-number-preference/create-account-number-preference.component.html]: Error for required field needs to be mentioned.

File: src/app/system/account-number-preferences/edit-account-number-preference/edit-account-number-preference.component.html [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+  
+      <form [formGroup]="accountNumberPreferenceForm" (ngSubmit)="submit()">
+  
+        <mat-card-content>
+  
+          <div fxLayout="column">
+  
+            <mat-form-field>
+                <mat-label>Account Type</mat-label>
+                <input matInput formControlName="accountType">
```
Comment (abhaychawla) [inline src/app/system/account-number-preferences/edit-account-number-preference/edit-account-number-preference.component.html]: Add `required`.

File: src/app/system/account-number-preferences/edit-account-number-preference/edit-account-number-preference.component.html [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+        <mat-card-actions fxLayout="row" fxLayout.xs="column" fxLayoutAlign="center" fxLayoutGap="5px">
+          <button type="button" mat-raised-button [routerLink]="['../']">Cancel</button>
+          <button mat-raised-button color="primary" [disabled]="!accountNumberPreferenceForm.valid || accountNumberPreferenceForm.pristine">Submit</button>
+        </mat-card-actions>
+  
+      </form>
+  
+    </mat-card>
+  
+  </div>
```
Comment (abhaychawla) [inline src/app/system/account-number-preferences/edit-account-number-preference/edit-account-number-preference.component.html]: Add a line at the end of file.

File: src/app/system/account-number-preferences/edit-account-number-preference/edit-account-number-preference.component.scss [hunk 12]
```diff
@@ -0,0 +1,3 @@
+.container {
+    max-width: 37rem;
+}
```
Comment (abhaychawla) [inline src/app/system/account-number-preferences/edit-account-number-preference/edit-account-number-preference.component.scss]: Add a line at the end of file.

File: src/app/system/account-number-preferences/edit-account-number-preference/edit-account-number-preference.component.ts [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+      this.accountNumberPreferencesTemplateData = data.accountnumberpreferencestemplate;
+    });
+  }
+
+  /**
+   * Creates and Sets account number preference form.
+   */
+  ngOnInit() {
+    this.createAccountNumberPreferenceForm();
+    this.tooglePrefixTypeStatus();
```
Comment (abhaychawla) [inline src/app/system/account-number-preferences/edit-account-number-preference/edit-account-number-preference.component.ts]:
```suggestion
    this.togglePrefixTypeStatus();
```

File: src/app/system/account-number-preferences/view-account-number-preference/view-account-number-preference.component.html [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+      Delete
+  </button>
+
+</div>
+
+<div class="container">
+
+  <mat-card>
+
+    <form [formGroup]="accountNumberPreferenceForm">
```
Comment (abhaychawla) [inline src/app/system/account-number-preferences/view-account-number-preference/view-account-number-preference.component.html]: Don't use forms in view layout. Should be similar to view layouts in `accounting` section.

File: src/app/system/account-number-preferences/view-account-number-preference/view-account-number-preference.component.html [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+
+        </div>
+
+      </mat-card-content>
+
+    </form>
+
+  </mat-card>
+
+</div>
```
Comment (abhaychawla) [inline src/app/system/account-number-preferences/view-account-number-preference/view-account-number-preference.component.html]: Add a line at the end of file.

File: src/app/system/account-number-preferences/view-account-number-preference/view-account-number-preference.component.scss [hunk 16]
```diff
@@ -0,0 +1,3 @@
+.container {
+    max-width: 37rem;
+}
```
Comment (abhaychawla) [inline src/app/system/account-number-preferences/view-account-number-preference/view-account-number-preference.component.scss]: Add a line at the end of file.

File: src/app/system/account-number-preferences/view-account-number-preference/view-account-number-preference.component.ts [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+   * Creates and sets the account number preference Form.
+   */
+  ngOnInit() {
+    this.createAccountNumberPreferenceForm();
+  }
+
+  /**
+   * Creates the view account number preference form.
+   */
+  createAccountNumberPreferenceForm() {
```
Comment (abhaychawla) [inline src/app/system/account-number-preferences/view-account-number-preference/view-account-number-preference.component.ts]: Form not needed here. Update accordingly.

File: src/app/system/system.service.ts [hunk 18]
```diff
@@ -99,4 +99,51 @@ export class SystemService {
     return this.http.put(`/configurations/${configurationId}`, configuration);
   }
 
+  /**
+   * @returns {Observable<any>} Fetches Account Number Preferences.
```
Comment (abhaychawla) [inline src/app/system/system.service.ts]:
```suggestion
   * @returns {Observable<any>} Account number preferences.
```
Make similar change in all cases.

File: src/app/system/account-number-preferences/create-account-number-preference/create-account-number-preference.component.ts (hunk 8 above)
Comment (luckyman20) [reply src/app/system/account-number-preferences/create-account-number-preference/create-account-number-preference.component.ts]: It is a checkbox and its value should not be passed so that's why I am disabling it.

Comment (abhaychawla) [reply src/app/system/account-number-preferences/create-account-number-preference/create-account-number-preference.component.ts]: Remove this checkbox as it is not required. Keep prefix as an optional field.

**Comment by abhaychawla** [comment]
Rebase once the comments are resolved.

File: src/app/system/account-number-preferences/create-account-number-preference/create-account-number-preference.component.ts (hunk 8 above)
Comment (abhaychawla) [reply src/app/system/account-number-preferences/create-account-number-preference/create-account-number-preference.component.ts]: Remove it from the edit component as well.

**Comment by abhaychawla** [comment]
Update your commit message and the screenshots as I mentioned above. Resolve the merge conflicts.

File: src/app/system/account-number-preferences/edit-account-number-preference/edit-account-number-preference.component.html [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+    <form [formGroup]="accountNumberPreferenceForm" (ngSubmit)="submit()">
+
+      <mat-card-content>
+
+        <div fxLayout="column">
+
+          <mat-form-field>
+            <mat-label>Account Type</mat-label>
+            <input matInput formControlName="accountType" required>
+          </mat-form-field>
```
Comment (abhaychawla) [inline src/app/system/account-number-preferences/edit-account-number-preference/edit-account-number-preference.component.html]: Don't modify the control type. Keep select instead of input. You can disable it. Also, keep the error message.

File: src/app/system/account-number-preferences/edit-account-number-preference/edit-account-number-preference.component.ts [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+  styleUrls: ['./edit-account-number-preference.component.scss']
+})
+export class EditAccountNumberPreferenceComponent implements OnInit {
+
+  /** Account Number Preference Form */
+  accountNumberPreferenceForm: FormGroup;
+  /** Account Number Preference Data */
+  accountNumberPreferenceData: any;
+  /** Account Number Preferences Template Data */
+  accountNumberPreferencesTemplateData: any;
```
Comment (abhaychawla) [inline src/app/system/account-number-preferences/edit-account-number-preference/edit-account-number-preference.component.ts]: Make all similar changes here as mentioned for create.

File: src/app/system/account-number-preferences/edit-account-number-preference/edit-account-number-preference.component.ts [hunk 21]
```diff
@@ -0,0 +1,88 @@
...
+    this.createAccountNumberPreferenceForm();
+  }
+
+  /**
+   * Sets Prefix type data.
+   * Creates and sets the edit account number preference form.
+   */
+  createAccountNumberPreferenceForm() {
+    let prefixTypeId = '';
+    this.prefixTypeData = this.accountNumberPreferencesTemplateData.prefixTypeOptions[this.accountNumberPreferenceData.accountType.code];
```
Comment (abhaychawla) [inline src/app/system/account-number-preferences/edit-account-number-preference/edit-account-number-preference.component.ts]: Keep this in a separate function or `ngoninit`.

File: src/app/system/account-number-preferences/edit-account-number-preference/edit-account-number-preference.component.ts [hunk 22]
```diff
@@ -0,0 +1,88 @@
...
+    if (this.accountNumberPreferenceData['prefixType'] !== undefined) {
+      for (const prefixTypeOption of this.prefixTypeData) {
```
Comment (abhaychawla) [inline src/app/system/account-number-preferences/edit-account-number-preference/edit-account-number-preference.component.ts]: You don't need to use a for loop here. Just directly assign `this.accountNumberPreferenceData.prefixType.id` while building the form.

File: src/app/system/system.module.ts [hunk 23]
```diff
@@ -54,7 +58,12 @@ import { EditCodeComponent } from './codes/edit-code/edit-code.component';
     EditAmazonS3Component,
     EditEmailComponent,
     EditNotificationComponent,
-    EditSMSComponent
+    EditSMSComponent,
+    EditConfigurationComponent,
```
Comment (abhaychawla) [inline src/app/system/system.module.ts]: This has been imported twice. Remove this line.

File: src/app/system/system-routing.module.ts [hunk 24]
```diff
@@ -273,7 +322,11 @@ const routes: Routes = [
     AmazonS3ConfigurationResolver,
     EmailConfigurationResolver,
     SMSConfigurationResolver,
-    NotificationConfigurationResolver
+    NotificationConfigurationResolver,
+    GlobalConfigurationResolver,
```
Comment (abhaychawla) [inline src/app/system/system-routing.module.ts]: This has been added twice. Remove this line.

File: src/app/system/system-routing.module.ts [hunk 25]
```diff
@@ -249,8 +256,50 @@ const routes: Routes = [
...
+              path: '',
+              component: AccountNumberPreferencesComponent,
+              resolve: {
+                accountNumberPreferences: AccountNumberPreferencesResolver
+              }
+            },
+            {
+              path: 'create',
+              component: CreateAccountNumberPreferenceComponent,
+              data: { title: extract('Create Account Number Preference'), breadcrumb: 'Create Account Number Preference' },
```
Comment (abhaychawla) [inline src/app/system/system-routing.module.ts]:
```suggestion
              data: { title: extract('Create Account Number Preference'), breadcrumb: 'Create' },
```

File: src/app/system/account-number-preferences/view-account-number-preference/view-account-number-preference.component.ts [hunk 26]
```diff
@@ -0,0 +1,61 @@
...
+    this.route.data.subscribe((data: { accountNumberPreference: any }) => {
+      this.accountNumberPreferenceData = data.accountNumberPreference;
+    });
+  }
+
+  ngOnInit() {
+  }
+
+  /**
+   * Deletes the current account number preference.
```
Comment (abhaychawla) [inline src/app/system/account-number-preferences/view-account-number-preference/view-account-number-preference.component.ts]:
```suggestion
   * Deletes the account number preference and redirects to account number preferences.
```

File: src/app/system/account-number-preferences/create-account-number-preference/create-account-number-preference.component.html [hunk 27]
```diff
@@ -0,0 +1,45 @@
...
+    <form [formGroup]="accountNumberPreferenceForm" (ngSubmit)="submit()">
+
+      <mat-card-content>
+
+        <div fxLayout="column">
+
+          <mat-form-field>
+            <mat-label>Select Account Type</mat-label>
+            <mat-select required formControlName="accountType">
+              <mat-option *ngFor="let accountType of accountNumberPreferencesTemplateData.accountTypeOptions" [value]="accountType.id">
```
[the remark this answers is not in the document]
Comment (luckyman20) [reply src/app/system/account-number-preferences/create-account-number-preference/create-account-number-preference.component.html]: Can I ask how will that help? Is there anything I am missing?

File: src/app/system/account-number-preferences/view-account-number-preference/view-account-number-preference.component.html [hunk 28]
```diff
@@ -0,0 +1,45 @@
...
+      Delete
+  </button>
+
+</div>
+
+<div class="container">
+
+  <mat-card>
+
+      <mat-card-content>
```
[the remark this answers is not in the document]
Comment (luckyman20) [reply src/app/system/account-number-preferences/view-account-number-preference/view-account-number-preference.component.html]: This one is correct or not? I will make the changes accordingly in rest of the files.

[the remark this answers is not in the document]
Comment (abhaychawla) [reply src/app/system/account-number-preferences/view-account-number-preference/view-account-number-preference.component.html]: For line 17 there are 2 spaces, then for line 19 there are 4. Its incorrect. It should be the same for all the lines.

File: src/app/system/account-number-preferences/create-account-number-preference/create-account-number-preference.component.html (hunk 27 above)
[the remark this answers is not in the document]
Comment (abhaychawla) [reply src/app/system/account-number-preferences/create-account-number-preference/create-account-number-preference.component.html]: It's just to follow a similar approach to what we have done in other components.

[the remark this answers is not in the document]
Comment (luckyman20) [reply src/app/system/account-number-preferences/create-account-number-preference/create-account-number-preference.component.html]: Duplicate review comment.

[the remark this answers is not in the document]
Comment (luckyman20) [reply src/app/system/account-number-preferences/create-account-number-preference/create-account-number-preference.component.html]: We are using `prefixTypeOptions` again (see line 57), so should I make two different variable for `accountTypeOptions` and `prefixTypeOptions`.

File: src/app/system/account-number-preferences/account-number-preferences.component.ts [hunk 29]
```diff
@@ -0,0 +1,64 @@
...
+    };
+    this.dataSource.sort = this.sort;
+  }
+
+  /**
+   * Filters data in account number preferences table based on passed value.
+   * @param {string} filterValue Value to filter data.
+   */
+  applyFilter(filterValue: string) {
+    this.dataSource.filter = filterValue.trim().toLowerCase();
```
Comment (abhaychawla) [inline src/app/system/account-number-preferences/account-number-preferences.component.ts]: The filter is not working as the data is in a nested object. You'll have to make a custom filter to make it work.

File: src/app/system/account-number-preferences/create-account-number-preference/create-account-number-preference.component.html (hunk 27 above)
[the remark this answers is not in the document]
Comment (abhaychawla) [reply src/app/system/account-number-preferences/create-account-number-preference/create-account-number-preference.component.html]: I see. Let us keep it as it is for this case. No need to make this change.

File: src/app/system/account-number-preferences/edit-account-number-preference/edit-account-number-preference.component.ts [hunk 30]
```diff
@@ -0,0 +1,79 @@
...
+    this.createAccountNumberPreferenceForm();
+  }
+
+  /**
+   * Creates and sets the edit account number preference form.
+   */
+  createAccountNumberPreferenceForm() {
+    this.accountNumberPreferenceForm = this.formBuilder.group({
+      'accountType': [{ value: this.accountNumberPreferenceData.accountType.id, disabled: true }, Validators.required],
+      'prefixType': [this.accountNumberPreferenceData.prefixType.id]
```
Comment (abhaychawla) [inline src/app/system/account-number-preferences/edit-account-number-preference/edit-account-number-preference.component.ts]: In case the prefix type is absent, it gives an error. Ensure that case is handled by first checking whether prefixType is present. If yes, then initialize with value otherwise with blank value. Use conditional operator.
